### PR TITLE
Include support for adding or setting response headers.

### PR DESCRIPTION
**Pull request by tobz**
The crucial bit here is that adding is additive only (i.e. you can double add a header) whereas setting will also add, but if it finds a header with the same name, it will simply update that existing header.

It will only override the first header it finds that matches, though, so if you double add a header, and then try to set the same header, it's only going to update the first.  I don't think that's a corner case worth dealing with, though, unless someone has a good use case to warrant scanning all headers outright.

### CODE REVIEW

**Comment by tobz** [comment]
Beep beep poke poke. :)

/cc @kellabyte

**Comment by kellabyte** [comment]
Do you think we should keep `http_response.http_header` as an array or should we use `khash` like I did elsewhere? I may have done this because it was quicker at the time. I may have justified it because adding new headers should mostly be additive most of the time and an array is fast with small sets.

**Comment by tobz** [comment]
I guess, if we're trying to be objective, it boils down to: how much do we care about performance?  how much do we care about readability?

We can do it both ways.  Maybe we should just do some performance testing here.  We can achieve the same exact functionality either way.  If the performance is close enough, then maybe we think about readability?

(fwiw, what's your current benchmarking setup, at a high level? OS, client vs server setup, etc)

**Comment by kellabyte** [comment]
Make sure you don't benchmark on localhost. Use 2 separate machines for client and server.

This script is an example of a benchmark you can run on the client against Haywire:
https://github.com/kellabyte/Haywire/blob/master/benchmark/benchmark.sh

You can change the arguments to Wrk to change different workload scenarios like more connections or less pipelined requests.

You probably want to benchmark against small amounts of headers vs a decent amount of headers and different workloads of mostly additive vs 50% additive and 50% update or something.

Bottom line, we care about performance in every change we make. My goal with Haywire is to try and be the fastest HTTP server there is. Lofty goals but it keeps the focus on performance in mind :)

**Comment by kellabyte** [comment]
I believe this is what nginx is using to store headers: https://github.com/nginx/nginx/blob/220c36d6128bcf3e869fdd0f7767ead9a4222f08/src/core/ngx_list.c

**Comment by tobz** [comment]
Super preliminary: I can't actually measure the difference between sending along one "extra" header (excluding content type) vs sending along an extra four.  So that's good, I guess. :)  At least I now have the baseline to benchmark a khash approach against.

Unfortunately, I can't get Haywire to actually scale to more than one core.  Some sort of weird... under load it besides to lock up/freeze up and only one worker thread can actually do anything.

(Linux 3.14, Amazon Linux, GCC 4.8; tried with 2 threads, 4, 8, etc.)

**Comment by kellabyte** [comment]
Is this lockup happening because of the new changes or without them? Is there a crash happening that you might be able to trap with GDB?

**Comment by tobz** [comment]
Didn't get that far.  I was testing with big instances courtesy of my previous employer, so now I need to wait until I have more than just my desktop to run anymore benchmarks, wahhhh. :(

Obviously most anyone can pick this up, just won't be me for another week or two.

**Comment by nmdguerreiro** [comment]
I benchmarked this PR on two m4.xlarge machines on AWS. Haywire was setup to use 4 threads.

### Current master
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.31ms  135.91us  13.23ms   95.82%
    Req/Sec   763.07     90.24     2.46k    97.99%
  Latency Distribution
     50%    1.32ms
     75%    1.35ms
     90%    1.38ms
     99%    1.46ms
  385766 requests in 5.10s, 58.13MB read
Requests/sec:  75647.78
Transfer/sec:     11.40MB
```

### @tobz's changes - `GET /``
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.19ms  154.72us  16.21ms   92.83%
    Req/Sec   835.31    104.96     2.27k    97.17%
  Latency Distribution
     50%    1.21ms
     75%    1.27ms
     90%    1.32ms
     99%    1.39ms
  385649 requests in 5.10s, 58.11MB read
  Socket errors: connect 0, read 5, write 0, timeout 0
  Non-2xx or 3xx responses: 175
Requests/sec:  75617.78
Transfer/sec:     11.39MB
```

### @tobz's changes - `GET /double`
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.16ms  205.87us  22.33ms   87.71%
    Req/Sec     0.86k   118.57     1.83k    84.24%
  Latency Distribution
     50%    1.18ms
     75%    1.27ms
     90%    1.32ms
     99%    1.40ms
  384622 requests in 5.10s, 74.26MB read
  Socket errors: connect 0, read 57, write 0, timeout 0
  Non-2xx or 3xx responses: 2060
Requests/sec:  75435.91
Transfer/sec:     14.56MB
```

### @tobz's changes - `GET /override`
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.12ms  161.15us  10.28ms   75.53%
    Req/Sec     0.88k   126.64     2.26k    78.61%
  Latency Distribution
     50%    1.11ms
     75%    1.24ms
     90%    1.32ms
     99%    1.40ms
  385123 requests in 5.10s, 74.10MB read
  Socket errors: connect 0, read 28, write 0, timeout 0
  Non-2xx or 3xx responses: 909
Requests/sec:  75514.61
Transfer/sec:     14.53MB
```

### @tobz's changes - `GET /mixandmatch`
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.26ms  134.63us  10.96ms   94.61%
    Req/Sec   793.59     82.68     1.96k    97.84%
  Latency Distribution
     50%    1.27ms
     75%    1.31ms
     90%    1.35ms
     99%    1.42ms
  384869 requests in 5.10s, 117.64MB read
  Socket errors: connect 0, read 28, write 0, timeout 0
  Non-2xx or 3xx responses: 1105
Requests/sec:  75470.10
Transfer/sec:     23.07MB
```

As you can see, results are pretty much identical, just that having more headers is a bit slower as is expected.

You'll see a number of non 2xx/3xx responses, but that is not caused by this commit, but by a race condition that exists on `route_compare_method.c`. I opened #78 to fix it.

**Comment by tobz** [comment]
@nmdguerreiro thanks for doing some consistent tests on this!

**Comment by nmdguerreiro** [comment]
No problem :sunglasses:

**Comment by kellabyte** [comment]
I did some more aggressive benchmarking.

# Master branch
```
../bin/wrk/wrk --script ./pipelined_get.lua --latency -d 10s -t 40 -c 512 http://server:8000 -- 32
Running 10s test @ http://server:8000
  40 threads and 512 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    16.10ms   12.74ms 101.92ms   55.55%
    Req/Sec    64.43k    20.61k  177.88k    70.60%
  Latency Distribution
     50%   14.12ms
     75%   27.06ms
     90%   34.87ms
     99%   41.82ms
  25786650 requests in 10.10s, 3.79GB read
Requests/sec: 2553250.20
Transfer/sec:    384.73MB
```

# This PR branch
```
./bin/wrk/wrk --script ./pipelined_get.lua --latency -d 10s -t 40 -c 512 http://server:8000 -- 32
Running 10s test @ http://server:8000
  40 threads and 512 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    18.74ms   12.09ms 119.86ms   57.91%
    Req/Sec    23.28k    11.42k   70.10k    67.40%
  Latency Distribution
     50%   18.58ms
     75%   29.01ms
     90%   35.31ms
     99%   39.87ms
  9006443 requests in 10.10s, 1.32GB read
  Socket errors: connect 0, read 29925, write 0, timeout 0
  Non-2xx or 3xx responses: 1642156
Requests/sec: 891678.30
Transfer/sec:    133.47MB
```

Should we test this with fix #78?

**Comment by nmdguerreiro** [comment]
I think I did but I haven't posted the results. I'll do that on Monday 😊

**Comment by nmdguerreiro** [comment]
Benchmarks on two m4.xlarge machines on AWS again with 4 server threads.

### Current master
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.36ms  708.80us  13.83ms   85.20%
    Req/Sec   760.38    308.75     1.96k    49.61%
  Latency Distribution
     50%    1.14ms
     75%    1.58ms
     90%    2.36ms
     99%    3.76ms
  385396 requests in 5.10s, 58.07MB read
Requests/sec:  75575.82
Transfer/sec:     11.39MB
```

### @tobz's changes - `GET /`
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.35ms  596.64us   8.01ms   84.10%
    Req/Sec   761.06    263.46     2.60k    85.85%
  Latency Distribution
     50%    1.32ms
     75%    1.35ms
     90%    1.43ms
     99%    3.85ms
  385266 requests in 5.10s, 58.05MB read
Requests/sec:  75551.17
Transfer/sec:     11.38MB
```

### @tobz's changes - `GET /double`
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.31ms  425.96us   9.79ms   78.58%
    Req/Sec   760.57    197.30     1.87k    51.07%
  Latency Distribution
     50%    1.13ms
     75%    1.61ms
     90%    1.95ms
     99%    2.55ms
  385143 requests in 5.10s, 74.19MB read
Requests/sec:  75519.88
Transfer/sec:     14.55MB
```

### @tobz's changes - `GET /override`
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.31ms  138.89us  13.07ms   92.71%
    Req/Sec   761.07     57.55     1.61k    97.66%
  Latency Distribution
     50%    1.31ms
     75%    1.35ms
     90%    1.40ms
     99%    1.55ms
  384989 requests in 5.10s, 74.17MB read
Requests/sec:  75488.38
Transfer/sec:     14.54MB
```

### @tobz's changes - `GET /mixandmatch`
```
[ec2-user@ip-172-31-5-244 wrk]$ ./wrk -c 100 -t 100 -d 5s -s pipelined_get.lua --latency http://172.31.5.65:8000
Running 5s test @ http://172.31.5.65:8000
  100 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.31ms  114.05us   4.22ms   94.75%
    Req/Sec   761.40     59.44     1.69k    97.73%
  Latency Distribution
     50%    1.32ms
     75%    1.35ms
     90%    1.38ms
     99%    1.46ms
  385005 requests in 5.10s, 117.86MB read
Requests/sec:  75496.58
Transfer/sec:     23.11MB
```

Similar results as before, so it looks safe to merge.
